File: modulo-01-fundamentos-arquitectura/codigo/src/main/java/pe/joedayz/microservicios/modulo01/infra/InMemoryOrderRepository.java
package pe.joedayz.microservicios.modulo01.infra;

import pe.joedayz.microservicios.modulo01.ddd.order.Order;
import pe.joedayz.microservicios.modulo01.ddd.order.OrderId;
import pe.joedayz.microservicios.modulo01.ddd.order.OrderRepository;
import pe.joedayz.microservicios.modulo01.tenant.TenantId;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

/**
 * ADAPTADOR en memoria del puerto {@link OrderRepository} (DIP en accion).
 *
 * <p>La clave incluye el tenant, de modo que un {@code findById} nunca puede devolver
 * el pedido de otro tenant aunque coincidiera el OrderId.
 */
public class InMemoryOrderRepository implements OrderRepository {

    private final Map<String, Order> store = new ConcurrentHashMap<>();

    @Override
    public Optional<Order> findById(TenantId tenantId, OrderId id) {
        return Optional.ofNullable(store.get(key(tenantId, id)));
    }

    @Override
    public void save(Order order) {
        store.put(key(order.tenantId(), order.id()), order);
    }

    private String key(TenantId tenantId, OrderId id) {
        return tenantId.value() + "::" + id.value();
    }
}

File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Infrastructure/InMemoryOrderRepository.cs
using System.Collections.Concurrent;
using JoeDayz.Microservicios.Modulo01.Ddd.Orders;
using JoeDayz.Microservicios.Modulo01.Tenant;

namespace JoeDayz.Microservicios.Modulo01.Infrastructure;

/// <summary>
/// Adaptador de persistencia en memoria. La clave incluye el tenant: es la forma mas
/// simple de aislamiento multi-tenant (equivalente a un filtro global en EF Core).
/// </summary>
public sealed class InMemoryOrderRepository : IOrderRepository
{
    private readonly ConcurrentDictionary<string, Order> _store = new();

    public Order? FindById(TenantId tenantId, OrderId id) =>
        _store.GetValueOrDefault(Key(tenantId, id));

    public void Save(Order order) => _store[Key(order.TenantId, order.Id)] = order;

    private static string Key(TenantId tenantId, OrderId id) => $"{tenantId.Value}::{id.Value}";
}
